Migrate from Manifest V2 to Manifest V3

sfllaw opened the pull request:
The Chrome web store is warning that this extension [“This extension may soon no longer be supported because it doesn't follow best practices for Chrome extensions.”](https://chromewebstore.google.com/detail/bookmark-search/hhmokalkpaiacdofbcddkogifepbaijk)

![Image](https://github.com/user-attachments/assets/8b74c1b5-010d-4275-b01c-c8d8c18af2f8)

This is likely because of Manifest V2 is deprecated and this extension needs to be [Migrate to Manifest V3](https://developer.chrome.com/docs/extensions/develop/migrate). 

Please review this PR commit at a time. I have split the commits into logical changes, with the tradeoff that the extension is broken until the penultimate commit.

I have taken the liberty to bump the major version, since this change is backwards incompatible. I have also added myself as a contributor, if you don’t mind.

Closes: #22

sfllaw committed:
Save options using storage.sync instead of localStorage

The localStorage API is no longer available to extensions in Manifest
V3, but the storage.sync API has the additional benefit of using
Chrome Sync to propagate these options across devices.

https://developer.chrome.com/docs/extensions/develop/ui/options-page

For s_automatchText and s_automatchUrl, the storage.local API is used
as an exact replacement for the localStorage API.

sfllaw committed:
chrome.notifications replaces webkitNotifications

webkitNotifications is unsupported in Manifest V3 and has been
replaced by the chrome.notifications API:

https://developer.chrome.com/docs/extensions/how-to/ui/notifications

sfllaw committed:
bookmarksToSuggestions breaks on bookmark folders

The built-in chrome.bookmarks.search can return a BookmarkTreeNode
that has a title but does not have the optional url. This happens when
the BookmarkTreeNode represents a bookmark folder.

https://developer.chrome.com/docs/extensions/reference/api/bookmarks#type-BookmarkTreeNode

Since bookmark folders are not useful in omnibox search results, this
patch filters them out.

sfllaw committed:
Version 2.0.0, since Manifest V3 is backwards incompatible

This completes the migration from Manifest V2 to Manifest V3:
https://developer.chrome.com/docs/extensions/develop/migrate

As of 2024-10-10, select Manifest V2 extensions are being disabled and
the Chrome Web Store recommends installing an alternate Manifest V3
extension. By updating to Manifest V3, this extension will avoid this
fate.
https://developer.chrome.com/docs/extensions/develop/migrate/mv2-deprecation-timeline

This patch also updates the changelogs and adds contributor sfllaw to
LICENSE.

sfllaw committed:
Run Extension Manifest Converter to convert to Manifest V3

https://github.com/GoogleChromeLabs/extension-manifest-converter

For this project, the tool automatically:
- Changed the manifest version number.
- Upgraded the "background" field in the manifest.
- Updated the content security policy.

The Extension Manifest Converter combined search_common.js and
background.js to make service_worker.js. But it has split it apart
again by using importScripts.

Also, background.persistent is no longer supported in Manifest V3:
https://developer.chrome.com/docs/extensions/develop/migrate/to-service-workers#update-bg-field

alvinhochun commented:
Thanks for the PR! I gave this a quick scan; even though I am not familiar with the changes from Manifest V3 the overall changes look fine (haven't tested it yet).

Just one main issue:
This seems to completely discard options previously stored in localStorage. Can we somehow migrate the previous options? I found [this message](https://groups.google.com/a/chromium.org/g/chromium-extensions/c/6qiMo0P-XS4/m/jiFmGCH0AQAJ) which seem to indicate that localStorage is accessible on pages, so can we perhaps migrate the options by opening the whatsnew page?

Nits (just ranting really, feel free to ignore):
- I don't like how one commit merged two scripts just to be split into two again in the other; would be a bit easier to review individual commits if they are squashed.
- Is the rename from `background.js` to `service_worker.js` really necessary? That makes diffing for changes from the full series harder.

alvinhochun commented:
I just tested it for a bit. Searching and opening URLs seems to work fine, so that's good.

However, JavaScript bookmarklets doesn't seem to work. First time attempting to run a bookmarklet would raise `Uncaught (in promise) ReferenceError: confirm is not defined` at
https://github.com/alvinhochun/chrome-bookmark-search/blob/65737474adc97344f64a61665c3ed4d1f7a058c4/src/background.js#L135
Which makes sense, since service workers are not supposed to have that. I guess it can show a notification instead? Or perhaps just open the options page directly and highlight the bookmarklet option?

But even after enabling the bookmarklet option, trying to run a bookmarklet would raise `Uncaught (in promise) Error: JavaScript URLs are not allowed in API based extension navigations. Use chrome.scripting.executeScript instead.`

Overall, the situation with bookmarklet seems a bit overcomplicated, since the extension also needs to make sure the script is executed in the correct security context, which I don't really know how to handle. Frankly I would be fine to just have bookmarklet support disabled for now (temporarily or permanently) and let the user know when they try to run a bookmarklet.

sfllaw commented:
> Just one main issue: This seems to completely discard options previously stored in localStorage. Can we somehow migrate the previous options? I found [this message](https://groups.google.com/a/chromium.org/g/chromium-extensions/c/6qiMo0P-XS4/m/jiFmGCH0AQAJ) which seem to indicate that localStorage is accessible on pages, so can we perhaps migrate the options by opening the whatsnew page?

I could look into seeing if we could launch a temporary tab if there was an upgrade and storage.sync is empty? A separate upgrade.html and tracking a previous version string might be helpful for this?

> * I don't like how one commit merged two scripts just to be split into two again in the other; would be a bit easier to review individual commits if they are squashed.

Sure, I can rebase this. I wanted to keep the migration tool’s run in its own commit, but I can rearrange its work.

> * Is the rename from `background.js` to `service_worker.js` really necessary? That makes diffing for changes from the full series harder.

It’s not necessary, no. Just conventional. I can leave the file where it was.

> First time attempting to run a bookmarklet would raise `Uncaught (in promise) ReferenceError: confirm is not defined` 

We could definitely create a notification for turning this on. We can show buttons in notifications now. This could be a tristate of "true", "", or "false". Where an actual "false" value means the user has turned this off.

> Overall, the situation with bookmarklet seems a bit overcomplicated, since the extension also needs to make sure the script is executed in the correct security context, which I don't really know how to handle. Frankly I would be fine to just have bookmarklet support disabled for now (temporarily or permanently) and let the user know when they try to run a bookmarklet.

Good catch! I don’t use bookmarklets, so didn’t notice this. Do you have an example that you’ve been using to test?

It looks like Chrome has done this on purpose: https://github.com/chromium/chromium/commit/2ad1e106e27b2338d9508e7b950aba96a91a5f4a

It also looks like the scripting.executeScript API no longer allows you to inject strings any more: https://developer.chrome.com/docs/extensions/reference/api/scripting#injected_code

That means we would have to do something like (untested):
```js
const bookmarklet = text.substr(14);
const tabs = await chrome.tabs.query({
	active: true,
	currentWindow: true,
	lastFocusedWindow: true
});
if (!tabs) {
	console.error("no active tab in the current window");
}
chrome.scripting.executeScript({
	target : {tabId : tabs[0].id},
	func : eval, // bookmarklets are script injection
	args : [ bookmarklet ],
})
```

alvinhochun commented:
> > Just one main issue: This seems to completely discard options previously stored in localStorage. Can we somehow migrate the previous options? I found [this message](https://groups.google.com/a/chromium.org/g/chromium-extensions/c/6qiMo0P-XS4/m/jiFmGCH0AQAJ) which seem to indicate that localStorage is accessible on pages, so can we perhaps migrate the options by opening the whatsnew page?
> 
> I could look into seeing if we could launch a temporary tab if there was an upgrade and storage.sync is empty? A separate upgrade.html and tracking a previous version string might be helpful for this?

Checking the version major upgraded from being lower than 2 and storage.sync being empty would be good.

> > * I don't like how one commit merged two scripts just to be split into two again in the other; would be a bit easier to review individual commits if they are squashed.
> 
> Sure, I can rebase this. I wanted to keep the migration tool’s run in its own commit, but I can rearrange its work.
> 
> > * Is the rename from `background.js` to `service_worker.js` really necessary? That makes diffing for changes from the full series harder.
> 
> It’s not necessary, no. Just conventional. I can leave the file where it was.

Thanks.

> > First time attempting to run a bookmarklet would raise `Uncaught (in promise) ReferenceError: confirm is not defined`
> 
> We could definitely create a notification for turning this on. We can show buttons in notifications now. This could be a tristate of "true", "", or "false". Where an actual "false" value means the user has turned this off.
> 
> > Overall, the situation with bookmarklet seems a bit overcomplicated, since the extension also needs to make sure the script is executed in the correct security context, which I don't really know how to handle. Frankly I would be fine to just have bookmarklet support disabled for now (temporarily or permanently) and let the user know when they try to run a bookmarklet.
> 
> Good catch! I don’t use bookmarklets, so didn’t notice this. Do you have an example that you’ve been using to test?

I used an ancient EXIF info bookmarklet which doesn't really work anymore, but if you run it on a page with an image element (Google homepage works) then it will submit the image URL to an HTTP page, otherwise it shows an alert:

```js
javascript:var form=document.createElement("form");form.action="http:"+"//exif.regex"+".info/exif.cgi";form.method="get";var O=document.createElement("input");O.type="hidden";O.name="b";O.value=3;form.appendChild(O);O=document.createElement("input");O.type="hidden";O.name="referer";O.value=document.location;form.appendChild(O);O=null;function f1(obj){if(obj.links.length>0){var R=/.(jpe?g|nef|raw|tiff)$/i;var%20count=0;var%20url;for(i=0;i%3Cobj.links.length;i++){if(R.test(obj.links[i].href)){count++;url=obj.links[i].href}}if(count==1){var%20O=document.createElement(%22input%22);O.type=%22hidden%22;O.name=%22url%22;O.value=url;return%20O;}}return%20null;}function%20f2(obj){if(obj.images.length==1){var%20O=obj.createElement(%22input%22);O.type=%22hidden%22;O.name=%22url%22;O.value=obj.images[0].src;return%20O;}return%20null;}function%20f3(obj){if(obj.images.length%3E=1){var%20img;var%20max=0;for(i=0;i%3Cobj.images.length;i++){size=obj.images[i].width*obj.images[i].height;if(size%3Emax){max=size;img=obj.images[i];}}return%20img;}return%20null;}function%20f4(obj){var%20img=f3(obj);if(img){O=obj.createElement(%22input%22);O.type=%22hidden%22;O.name=%22url%22;O.value=img.src;return%20O;}return%20null;}function%20f5(FRAMES){for(var%20fr=0;O==null&&fr%3CFRAMES.length;fr++){try{var%20O=f1(FRAMES[fr].document);if(O)return%20O;}catch(er){}}var%20max=0;var%20img;for(var%20fr=0;O==null&&fr%3CFRAMES.length;fr++){try{var%20I=f3(FRAMES.document);if(I&&I.width*I.height%3Emax){img=I;max=I.width*I.height;}}catch(er){}if(img){O=obj.createElement(%22input%22);O.type=%22hidden%22;O.name=%22url%22;O.value=img.src;return%20O;}}return%20null}var%20O=f1(document);if(!O)O=f2(document);if(!O)O=f4(document);if(!O)O=f5(frames);if(O){form.appendChild(O);document.body.appendChild(form);form.submit();}else{alert(%22Sorry,I%20could%20not%20find%20out%20an%20image%20to%20report%20on%22);}
```

> It looks like Chrome has done this on purpose: [chromium/chromium@2ad1e10](https://github.com/chromium/chromium/commit/2ad1e106e27b2338d9508e7b950aba96a91a5f4a)
> 
> It also looks like the scripting.executeScript API no longer allows you to inject strings any more: https://developer.chrome.com/docs/extensions/reference/api/scripting#injected_code
> 
> That means we would have to do something like (untested):
> 
> ```js
> const bookmarklet = text.substr(14);
> const tabs = await chrome.tabs.query({
> 	active: true,
> 	currentWindow: true,
> 	lastFocusedWindow: true
> });
> if (!tabs) {
> 	console.error("no active tab in the current window");
> }
> chrome.scripting.executeScript({
> 	target : {tabId : tabs[0].id},
> 	func : eval, // bookmarklets are script injection
> 	args : [ bookmarklet ],
> })
> ```

If they block string injection I suspect `eval` might not work either, and I think `eval` can be disabled by the site using CSP headers. Also, if `chrome.scripting.executeScript` runs the script in the extension's context then we don't want that.

And given that the Chrome web store is tightening on extension permissions, I have a feeling maybe we should just drop bookmarklet support.

Speaking of which, I logged in the developer dashboard and found that I have to fill out some Privacy info which includes justifying for the permission usage. I've submitted a review but I have suspicion that my justification for the "activeTab" permission may be rejected.

sfllaw commented:
> If they block string injection I suspect `eval` might not work either, and I think `eval` can be disabled by the site using CSP headers. Also, if `chrome.scripting.executeScript` runs the script in the extension's context then we don't want that.
> 
> And given that the Chrome web store is tightening on extension permissions, I have a feeling maybe we should just drop bookmarklet support.

I looked into it and Firefox has been enforcing CSP on their bookmarklets for 12 years now: [#866522: Bookmarklets affected by CSP](https://bugzilla.mozilla.org/show_bug.cgi?id=866522). 

So yeah, I think dropping bookmarklet support is really the only sensible option for this transition.

sfllaw committed:
Remove jsbm support for `javascript:` bookmarklets

In Manifest V3, Chrome no longer allows extensions to run bookmarklets
with the following error:

    Uncaught (in promise) Error: JavaScript URLs are not allowed in
	API based extension navigations. Use
	chrome.scripting.executeScript instead.

In theory, using the scripting.executeScript API to eval the bookmarks
would be possible, but many sites have a Content Security Policy that
would block this behavior.

And the CSP implementation in Chrome even blocks legitimate
bookmarklets: https://issues.chromium.org/issues/40077444

So instead of providing a degraded experience, bookmarklets will no
longer suggested in the omnibox, even when their title or URL would
match.

sfllaw commented:
@alvinhochun: I have revised this patchset to address the following:
- [X] af8d928b930464ade323948285da500c0155ed21: I have minimized the differences generated by the Extension Manifest Converter. This undoes the renaming and merging of background.js.
- [X] 1ce3cca281faebf485f4aefda6d161f1d49eb7da: Removed bookmarklet support, since Chrome puts priority on Content-Security-Policy: script-src

On my TODO list is:
- [ ] converting the `window.confirm` call to a notification
- [ ] automatically migrating the options in `localStorage` to `storage.sync` from v1 to v2.

sfllaw committed:
Move options from localStorage to storage.sync when upgrading

In Manifest V3, extension service workers can't use
window.localStorage because they don't have access to the window
object.

https://developer.chrome.com/docs/extensions/develop/migrate/to-service-workers#convert-localstorage

Options used to be saved in localStorage, but saving them with
chrome.storage.sync means that all synced browsers will share the same
options via Chrome Sync.

https://developer.chrome.com/docs/extensions/reference/api/storage#storage_areas

After every upgrade, upgrade.html is opened in the background to
perform the data migration. The code in upgrade.js needs to run in a
real tab so it has access to window.localStorage. After it copies the
relevant options to chrome.storage.sync, it removes them from
localStorage.

sfllaw commented:
@alvinhochun: I think I have addressed all your concerns above:

* [x]  [af8d928](https://github.com/alvinhochun/chrome-bookmark-search/commit/af8d928b930464ade323948285da500c0155ed21): I have minimized the differences generated by the Extension Manifest Converter. This undoes the renaming and merging of background.js.
* [x]  6bef1ffa5fb8cb99b7147fa223d792713d01db2b: Removed bookmarklet support, since Chrome puts priority on Content-Security-Policy: script-src
* [x]  removing the `window.confirm` call, since it is about bookmarklets
* [x]  40feacc2a5c1ede4ca120a16a9763535024c95d9: automatically migrating the options in `localStorage` to `storage.sync` from v1 to v2.

alvinhochun commented on src/search_common.js:
```diff
@@ -66,65 +65,42 @@ var bookmarks = (function(){
+	var m = parseInt(options["maxcount"]);
+	for(let i = 0; s.length < m && i < b.length; i++){
 		var v = b[i];
+		if(!v.url || jsMatch.test(v.url)){
+			// javascript: bookmarklets are no longer supported,
+			// because modern websites have broken them with
+			// Content-Security-Policy: script-src
+			// https://issues.chromium.org/issues/40077444
+			continue;
+		}
```
This method of skipping results causes a logic error. If you look further down in `searchInput`:
https://github.com/alvinhochun/chrome-bookmark-search/blob/a63793a59cbb2aab62dcbe56f861ba46bb7848cf/src/search_common.js#L126-L174
The code expects the first element (index 0) of `b` (`results`) and `s` to refer to the same bookmark. If the first result matches the condition and is skipped, then that logic produces the wrong first result.

sfllaw replied:
Ah… that is a subtle bug. Good catch!

9839f6f17ce572a0806cfda89659931a40d488dc fixes this by tagging `s` with each associated `bookmark`, instead of trying to keep `b` and `s` in sync. I considered making `s` a list of tuples, but I think annotating and then stripping out the `.bookmark` is more obvious.

alvinhochun commented:
Thanks for the work! The migration seems to work perfectly. There is however a blocking issue with the logic of skipping results.

sfllaw committed:
`bookmarksToSuggestions` tags each suggestion with the bookmark

`bookmarksToSuggestions` generates a suggestion for every bookmark.
But since it ignores bookmark folders and bookmarklets, `s.length` may
be shorter than `b.length`.

`searchInput` assumed that there was a 1:1 match between every
bookmark and suggestion. Now, each suggestion is tagged with a
`bookmark` property that points to the bookmark it was generated from.

Unfortunately, `chrome.omnibox.sendSuggestions` rejects unexpected
properties, so it needs to `clean` the `bookmark` property from the
suggestions before `suggest` is called.

alvinhochun reviewed (approved):
Thanks, LGTM.
